### PR TITLE
Daily Slack Reports

### PR DESCRIPTION
**Pull request by Math3v**

### CODE REVIEW

File: lib/reddex/slack_rtm.ex [hunk 1]
```diff
@@ -0,0 +1,17 @@
+defmodule Reddex.SlackRtm do
```
Comment (Math3v) [inline lib/reddex/slack_rtm.ex]: Modules should have a @moduledoc tag.

File: lib/reddex/slack.ex [hunk 2]
```diff
@@ -0,0 +1,19 @@
+defmodule Reddex.Slack do
```
Comment (Math3v) [inline lib/reddex/slack.ex]: Modules should have a @moduledoc tag.

**Comment by Math3v** [comment]
Ebert has finished reviewing this Pull Request and has found:

* 2 possible new issues (including those that may have been commented here).

You can see more details about this review at https://ebertapp.io/github/PrimeHammer/reddex/pulls/22.

File: test/support/factories/link_factory.ex [hunk 3]
```diff
@@ -0,0 +1,10 @@
+defmodule Reddex.LinkFactory do
```
Comment (Math3v) [inline test/support/factories/link_factory.ex]: Modules should have a @moduledoc tag.

**Comment by Math3v** [comment]
Ebert has finished reviewing this Pull Request and has found:

* 1 possible new issue (including those that may have been commented here).

But beware that this branch is **[5 commits behind the `PrimeHammer:master` branch](https://github.com/PrimeHammer/reddex/compare/PrimeHammer:daily-report...PrimeHammer:master)**, and a review of an up to date branch would produce more accurate results.

You can see more details about this review at https://ebertapp.io/github/PrimeHammer/reddex/pulls/22.

File: lib/reddex/links/links.ex [hunk 4]
```diff
@@ -21,6 +21,14 @@ defmodule Reddex.Links do
     Repo.all(Link)
   end
 
+  def list_pending_links do
+    from(
+      l in Link,
+      where: is_nil(l.sent_to_slack_at)
+    )
+    |> Repo.all()
```
Comment (Math3v) [inline lib/reddex/links/links.ex]: Use a function call when a pipeline is only one function long

**Comment by Math3v** [comment]
Ebert has finished reviewing this Pull Request and has found:

* 2 possible new issues (including those that may have been commented here).

But beware that this branch is **[5 commits behind the `PrimeHammer:master` branch](https://github.com/PrimeHammer/reddex/compare/PrimeHammer:daily-report...PrimeHammer:master)**, and a review of an up to date branch would produce more accurate results.

You can see more details about this review at https://ebertapp.io/github/PrimeHammer/reddex/pulls/22.

File: test/reddex/reports/report_test.exs [hunk 5]
```diff
@@ -0,0 +1,18 @@
+defmodule Reddex.ReportTest do
+  use Reddex.DataCase
+
+  alias Reddex.{Report, ReportItem}
```
Comment (Math3v) [inline test/reddex/reports/report_test.exs]: Most of the time you are using the multiple single line alias/require/import/use directives but here you are using the multi-alias/require/import/use syntax

File: lib/reddex/reports/report.ex [hunk 6]
```diff
@@ -0,0 +1,13 @@
+defmodule Reddex.Report do
+  alias Reddex.Links
+  alias Reddex.ReportItem
+
+  def generate do
+    Links.list_pending_links()
+    |> Enum.map(&link_to_report_item/1)
```
Comment (Math3v) [inline lib/reddex/reports/report.ex]: Use a function call when a pipeline is only one function long

File: lib/reddex/reports/report.ex [hunk 7]
```diff
@@ -0,0 +1,13 @@
+defmodule Reddex.Report do
```
Comment (Math3v) [inline lib/reddex/reports/report.ex]: Modules should have a @moduledoc tag.

**Comment by Math3v** [comment]
Ebert has finished reviewing this Pull Request and has found:

* 3 possible new issues (including those that may have been commented here).

But beware that this branch is **[5 commits behind the `PrimeHammer:master` branch](https://github.com/PrimeHammer/reddex/compare/PrimeHammer:daily-report...PrimeHammer:master)**, and a review of an up to date branch would produce more accurate results.

You can see more details about this review at https://ebertapp.io/github/PrimeHammer/reddex/pulls/22.

File: lib/reddex/reports/report.ex [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+    Links.list_pending_links()
+    |> Enum.map(&link_to_report_item/1)
+  end
+
+  defp link_to_report_item(%{tags: tags} = link) do
+    %ReportItem{slack_message: slack_message(link), slack_channels: tags}
+  end
+
+  defp slack_message(%{id: id, title: title, description: description}) do
+    # TODO: Full URL to Reddex
```
Comment (Math3v) [inline lib/reddex/reports/report.ex]: TODO found

**Comment by Math3v** [comment]
Ebert has finished reviewing this Pull Request and has found:

* 4 possible new issues (including those that may have been commented here).

But beware that this branch is **[5 commits behind the `PrimeHammer:master` branch](https://github.com/PrimeHammer/reddex/compare/PrimeHammer:daily-report...PrimeHammer:master)**, and a review of an up to date branch would produce more accurate results.

You can see more details about this review at https://ebertapp.io/github/PrimeHammer/reddex/pulls/22.

File: lib/reddex/reports/report.ex [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+      Reddex.Links.mark_as_sent(link_id)
+    end)
+  end
+
+  defp link_to_report_item(%{tags: tags} = link) do
+    %ReportItem{link_id: link.id, slack_message: slack_message(link), slack_channels: tags}
+  end
+
+  defp slack_message(%{id: id, title: title, description: description}) do
+    # TODO: Full URL to Reddex
```
Comment (Math3v) [inline lib/reddex/reports/report.ex]: TODO found

File: lib/reddex/reports/report.ex [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+  end
+
+  defp send_report_item(%ReportItem{
+         link_id: link_id,
+         slack_message: slack_message,
+         slack_channels: slack_channels
+       }) do
+    slack_channels
+    |> Enum.each(fn channel ->
+      # TODO: Use channel variable instead of "#reddex"
```
Comment (Math3v) [inline lib/reddex/reports/report.ex]: TODO found

File: lib/reddex/scheduler.ex [hunk 11]
```diff
@@ -0,0 +1,4 @@
+defmodule Reddex.Scheduler do
```
Comment (Math3v) [inline lib/reddex/scheduler.ex]: Modules should have a @moduledoc tag.

File: lib/reddex/reports/report.ex [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+    |> Enum.map(&link_to_report_item/1)
+  end
+
+  defp send_report_item(%ReportItem{
+         link_id: link_id,
+         slack_message: slack_message,
+         slack_channels: slack_channels
+       }) do
+    slack_channels
+    |> Enum.each(fn channel ->
```
Comment (Math3v) [inline lib/reddex/reports/report.ex]: Use a function call when a pipeline is only one function long

File: lib/reddex/reports/report.ex [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+  alias Reddex.ReportItem
+
+  def send do
+    generate()
+    |> Enum.each(&send_report_item/1)
+  end
+
+  def generate do
+    Links.list_pending_links()
+    |> Enum.map(&link_to_report_item/1)
```
Comment (Math3v) [inline lib/reddex/reports/report.ex]: Use a function call when a pipeline is only one function long

File: lib/reddex/links/link.ex [hunk 14]
```diff
@@ -23,6 +24,11 @@ defmodule Reddex.Links.Link do
     |> tags_input_to_tags_array()
   end
 
+  def update_changeset(link, attrs) do
+    link
+    |> cast(attrs, [:title, :description, :sent_to_slack_at])
```
Comment (Math3v) [inline lib/reddex/links/link.ex]: Use a function call when a pipeline is only one function long

**Comment by Math3v** [comment]
Ebert has finished reviewing this Pull Request and has found:

* 9 possible new issues (including those that may have been commented here).

But beware that this branch is **[5 commits behind the `PrimeHammer:master` branch](https://github.com/PrimeHammer/reddex/compare/PrimeHammer:daily-report...PrimeHammer:master)**, and a review of an up to date branch would produce more accurate results.

You can see more details about this review at https://ebertapp.io/github/PrimeHammer/reddex/pulls/22.

File: lib/reddex/reports/report.ex [hunk 15]
```diff
@@ -0,0 +1,36 @@
+defmodule Reddex.Report do
+  alias Reddex.Links
+  alias Reddex.ReportItem
+
+  def send do
+    generate()
+    |> Enum.each(&send_report_item/1)
+  end
+
+  def generate do
```
Comment (erich) [inline lib/reddex/reports/report.ex]: I would rename this function, because you don't generate anything here. You select and format the links.

**Comment by Math3v** [comment]
Ebert has finished reviewing this Pull Request and has found:

* 8 possible new issues (including those that may have been commented here).

But beware that this branch is **[5 commits behind the `PrimeHammer:master` branch](https://github.com/PrimeHammer/reddex/compare/PrimeHammer:daily-report...PrimeHammer:master)**, and a review of an up to date branch would produce more accurate results.

You can see more details about this review at https://ebertapp.io/github/PrimeHammer/reddex/pulls/22.

File: lib/reddex/reports/report.ex [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+    |> Enum.map(&link_to_report_item/1)
+  end
+
+  defp send_report_item(%ReportItem{
+         link_id: link_id,
+         slack_message: slack_message,
+         slack_channels: slack_channels
+       }) do
+    Enum.each(slack_channels, fn _channel ->
+      # TODO: Use channel variable instead of "#reddex"
```
Comment (Math3v) [inline lib/reddex/reports/report.ex]: TODO found

File: lib/reddex/reports/report.ex [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+  alias Reddex.ReportItem
+
+  def send do
+    get_report_items()
+    |> Enum.each(&send_report_item/1)
+  end
+
+  def get_report_items do
+    Links.list_pending_links()
+    |> Enum.map(&link_to_report_item/1)
```
Comment (Math3v) [inline lib/reddex/reports/report.ex]: Use a function call when a pipeline is only one function long

File: lib/reddex/reports/report.ex [hunk 18]
```diff
@@ -0,0 +1,37 @@
+defmodule Reddex.Report do
+  @moduledoc false
+
+  alias Reddex.Links
+  alias Reddex.ReportItem
+
+  def send do
+    get_report_items()
+    |> Enum.each(&send_report_item/1)
```
Comment (Math3v) [inline lib/reddex/reports/report.ex]: Use a function call when a pipeline is only one function long

**Comment by Math3v** [comment]
Ebert has finished reviewing this Pull Request and has found:

* 3 possible new issues (including those that may have been commented here).

But beware that this branch is **[5 commits behind the `PrimeHammer:master` branch](https://github.com/PrimeHammer/reddex/compare/PrimeHammer:daily-report...PrimeHammer:master)**, and a review of an up to date branch would produce more accurate results.

You can see more details about this review at https://ebertapp.io/github/PrimeHammer/reddex/pulls/22.

**Comment by Math3v** [comment]
Ebert has finished reviewing this Pull Request and has found:

* 3 possible new issues (including those that may have been commented here).

You can see more details about this review at https://ebertapp.io/github/PrimeHammer/reddex/pulls/22.

**Comment by Math3v** [comment]
Ebert has finished reviewing this Pull Request and has found:

* 1 possible new issue (including those that may have been commented here).

You can see more details about this review at https://ebertapp.io/github/PrimeHammer/reddex/pulls/22.

**Pull request merged**